### ISSUE TITLE
Design Idea: Minimal OpenJFX UI

### ISSUE THREAD

**Issue by Urwumpe** [mars-sim/mars-sim#187]
Since we have more issues with legacy UI libraries than with libraries used in the core, what about adding an alternative JavaFX UI there, aiming at reducing the risks of outdated dependencies. I know that this could mean inventing the wheel twice, but would allow a more stable base of development. 

Technically, it should be only use Mars-sim or dependencies used in mars-sim-core as dependency. Also it should not make use of Swing in this case. To keep the effort low, the UI should be minimalist. No fancy animations, no complex drag and drop magic, just a clean and simple UI, that could maybe require a few clicks more to see some details. 

Instead of the current multiwindow approach, I would suggest using a paged approach, maybe like that: 

- Left side: Categories or Views
- Center: Content
- Right side Top: Mars Mini-Map or settlement mini-Map (Maybe also "Site mini-map" for Missions?)
- Right side bottom: Bookmarks (e.g. "Medical status of Person XYZ")
- Top row: Time and global Mars status or Time and settlement status, at the right edge of it a menu with configuration options and a in-game help.
- Row 2:  Shortcut icons to important personal views in commander mode. 
- Bottom, upper: Navigation bar (e.g. "Alpha Base > Greenhouse Status > Greenhouse 1") 
- Bottom, lower: Messages 

Commander mode should be turn based, in the sense that the simulation waits for action of the player when the commander has nothing to do. 

Instead of the big detailled dialogs we have now, it should use a layered approach: It starts at the most abstract level and further clicks on the content allow to reach finer levels with more detailled data. Bookmarks should make it easier to reach the most detailled levels for the player.

It should be possible to customize the UI towards the needs of the player, but for keeping the need for other dependencies minimal, this should be a secondary goal. Should a cuter version of this UI be needed, it should be done by forking and including new dependencies there, instead of adding new dependencies here.

One fanciness I could imagine adding to this UI already would be a 3D view of the base, as long as there is a common basic module in mars-sim to provide the models. A similar module would be useful anyway to serve the current 2D graphics of the base. 

What do you think?

**Comment by mokun** [comment]
Sure! The current graphic UI is due for overhaul.

Having said that, do we want to use OpenJFX API? 

We do have a number of choice such as nifty, JOGL, LWJGL, FXGL, etc. ?

I looked up some LWJGL videos on a youtube channel called ThinMatrix. He uses LWJGL to create a cute game called [Equilinox](https://www.youtube.com/watch?v=XuTXzLZMLUE). Of course, LWJGL uses  OpenGL and has great potential in pushing the graphical envelope.

Currently I'm taking a look at some examples using [nifty](https://github.com/nifty-gui/nifty-gui
) with its java2d renderer. 

In the past, I have thought of using FXGL as it's the only one that runs on top of javafx API and its uses some Kotlin.

**Comment by Urwumpe** [comment]
Well, I think JavaFX would still be the better compromise there for keeping the effort low. Also it would have a rather modest hardware footprint compared to better game engines. But that does not mean that there can be only one - there could still be more sophisticated UIs around. I just wanted to have something like a "minimum viable UI" beyond the headless mode. 

Especially for better keeping track with the changes in the core. When more effort is needed to modify the UI, it means a longer delay until a feature inside the core is supported by the UI.

**Comment by mokun** [comment]
Yes to "minimum viable UI". There's going to be a learning curve on something new. OpenJFX is probably the best to go for simulation genre. 

I've played with some LWJGL demos

I just came across this art demo called [Project Eagle](https://store.steampowered.com/app/224000/Project_Eagle_A_3D_Interactive_Mars_Base/) this afternoon in Steam.

**Comment by mokun** [comment]
We could say that our Java Swing will remain as the minimum viable UI. It works in Windows and Mac now. I suppose it's easier to make it work in Linux under Java 11 since Linux may use different versions of GTKs under Java 8. 

I just wanna go beyond the spreadsheet look but yes it needs achievable in near terms.

**Comment by mokun** [comment]
Alternatively, we could develop an [Open MCT plug-in](https://github.com/nasa/openmct) for mars-sim and spend more time on refining the core engine.

**Comment by Urwumpe** [comment]
Not sure about if Open MCT works for us as UI - I see some big gaps there. 

But its a great find for another project I know... I'll suggest it there. 

I thought about using a simple RCP framework at a point, but I think the idea should be keeping the dependencies low... swing is better for achieving that goal, but JFX looks better and allows more artistic freedom with CSS.

**Comment by shirishag75** [comment]
What should be looked at is how we can have something which can be used both on the mobile and the desktop space without sacrificing much functionality and have some sort of syncing. Wouldn't it be great if we could have the sim. run on both at times and sync or/and jump from one or the other or even have two different simulations with two or more different settings ensuring different game endings or stories.  We could also have a story-mode after the game has ended highlighting what the gamer did wrong or right and giving him hints (begineer hints.) but that's another story altogether I guess.

**Comment by mokun** [comment]
@shirishag75 , 

> can be used both on the mobile and the desktop space without sacrificing much functionality and have some sort of syncing. 

Yes this is a platform question that we can discuss in detail on a separate thread.

> wouldn't it be great if we could have the sim. run on both at times and sync or/and jump from one or the other or even have two different simulations with two or more different settings ensuring different game endings or stories. 

This is what they call "Distributed Architecture for Interactive Multiplayer Games

> We could also have a story-mode after the game has ended highlighting what the gamer did wrong or right and giving him hints (begineer hints.) 

I would love to have some sort of story-mode in mars-sim. It's a matter of conceptualizing what it can do. Will certainly make things more alive in the simulation :D

**Comment by mokun** [comment]
@Urwumpe, 

We could adopt LibGDX as our game engine as it also uses LWJGL to further provide for low level bindings that give access to many other useful APIs such as OpenGL, OpenCL, OpenAL, GLFW, CUDA lib, etc.. This opens up the possibility of porting mars-sim to AR/VR platforms such as Oculus and Magic Leap, etc. 

This way, our development will get more interesting and more people may join in.
